### ISSUE TITLE
Issues with notFilter

### ISSUE THREAD

**Issue by andreif** [cannatag/ldap3#255]
If one of `NamedTypes().__namedTypes` is `notFilter`, then `getTagMap()` fails due to `nt.getType().getTagMap()` being `None`. Is it `pyasn1`'s issue or `ldap3`'s one?

`nt.getType()` returns `ldap3.protocol.rfc4511.Not` instance which is impossible to print with Python 3.5:
```py
  File "/Users/andrei/.pyenv/versions/ldap/lib/python3.5/site-packages/pyasn1/type/namedtype.py", line 148, in getTagMap
    print(nt.getType())
  File "/Users/andrei/.pyenv/versions/ldap/lib/python3.5/site-packages/pyasn1/type/univ.py", line 452, in __str__
    def __str__(self): return self._value.decode(self._encoding, 'ignore')
  File "/Users/andrei/.pyenv/versions/ldap/lib/python3.5/site-packages/pyasn1/type/base.py", line 57, in __getattr__
    raise error.PyAsn1Error('No value for %s()' % attr)
pyasn1.error.PyAsn1Error: No value for decode()
```

#### Update:
`Not().getTagMap` is inherited from `Choice` and runs `Set.getComponentTagMap(self)` (due to empty `Not()._tagSet`):
```py
    def getComponentTagMap(self):
        if self._componentType:
            return self._componentType.getTagMap(True)
```
where `Not()._componentType` is empty so it returns `None` and fails later in various places.

**Comment by cannatag** [comment]
Hi,I'm not sure what the problem is. Can you give me some example code? You shouldn't need to interact with the asn.1 definition in ldap3. What is your use case?

**Comment by andreif** [comment]
@cannatag I am testing [devise_ldap_authenticatable](https://github.com/cschiewek/devise_ldap_authenticatable) for Ruby-on-Rails using a `ldap3`-based server. The Rails app sends the first bind message, on which I respond with success. Then I get the second message which I try to `decode(message, asn1Spec=LDAPMessage())` and get the error. The second message is:
```py
b'0\x81\xaa\x02\x01\x02c\x81\xa4\x04\x00\n\x01\x00\n\x01\x00\x02\x01\x00\x02\x01\x00\x01\x01\x00\x87\x0bobjectclass0\x81\x83\x04\x0enamingContexts\x04\x14supportedLdapVersion\x04\taltServer\x04\x10supportedControl\x04\x12supportedExtension\x04\x11supportedFeatures\x04\x17supportedSASLMechanisms'
```
Disclaimer: I am very new to LDAP and likely doing it wrong.

**Comment by cannatag** [comment]
Thatvseems a packet requesting info on the ldap server. It's quite complex to understand if you're new to ldap. You should decode it and then return the requested information. 

By the way what do you mean  with  "ldap3 based server"?

**Comment by andreif** [comment]
The second message seems to be `DsaInfo` as I see now.

I needed to mock a LDAP server to test authentication flow in the application without changing the app code.

**Comment by cannatag** [comment]
So are you using ldap3 to decode the messages sent to an ldap server? Even if  Ldap3 is a client ldap library it includes all the asn.1 definitions used by the server. But I'm not sure if they get properly decoded. 

Decoding an ldap filter is quite complex because the filter is not sent as a string but as a recursive asn.1 structure, this could be the reason for your error. If you send me your source code I will try to check if  it works correctly.

**Comment by andreif** [comment]
Yes, I find it very useful and use mostly protocols. In my case I just need to respond with minimal info (basically `ok`) so that the user can log in. I hope that everything I need is already in `ldap3` and I just need to learn and understand how it works in order to solve it.

**Comment by andreif** [comment]
It works fine when I add fallback value to `Set`:
```py
class Set(SequenceAndSetBase):
    ...
    def getComponentTagMap(self):
        if self._componentType:
            return self._componentType.getTagMap(True)
        return tagmap.TagMap()  # ADDED
    ...
```

**Comment by andreif** [comment]
So now I can log in using my dummy LDAP server, sweet. Thanks for your nice and clean library.

**Comment by cannatag** [comment]
You're welcome. Thanks for using ldap3.

Giovanni

**Comment by andreif** [comment]
Let's see if it will be fixed via etingof/pyasn1#12
